Modules/_asynciomodule.c

```diff
@@ -1601,11 +1601,24 @@ static void
 FutureIter_dealloc(futureiterobject *it)
 {
     PyTypeObject *tp = Py_TYPE(it);
-    asyncio_state *state = get_asyncio_state_by_def((PyObject *)it);
+
+    assert(_PyType_HasFeature((PyTypeObject *)tp, Py_TPFLAGS_HEAPTYPE));
+
+    PyHeapTypeObject *ht = (PyHeapTypeObject*)tp;
+    PyObject *module = ht->ht_module;
+    asyncio_state *state = NULL;
+
     PyObject_GC_UnTrack(it);
     tp->tp_clear((PyObject *)it);
 
-    if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
+    // GH-115874: We can't use PyType_GetModuleByDef here as the type might have
+    // already been cleared. This is also why we must check if ht_module != NULL.
+    // Subclasses also cannot make use of the free list.
+    if (module && _PyModule_GetDef(module) == &_asynciomodule) {
```

Comment:
Another thought: the control-flow here might be a bit clearer if rewritten. I'm not sure if the early return makes things more or less complex, though, so I'll let you make the call.

```c
    if (module && _PyModule_GetDef(module) == &_asynciomodule) {
        state = get_asyncio_state(module);
        if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
            state->fi_freelist_len++;
            it->future = (FutureObj *)state->fi_freelist;
            state->fi_freelist = it;
            return;
        }
    }
    PyObject_GC_Del(it);
    Py_DECREF(tp);
```

(GH won't let me make suggestions outside of the diff.)

Reply:
Personally, I probably think the way it's written now it's a bit easier to parse so I'll leave it as is unless someone else feels strongly.